storage_test.go

```diff
@@ -0,0 +1,442 @@
+package dynamodbstorage
+
+import (
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/caddyserver/certmagic"
+)
+
+const TestTableName = "CertMagicTest"
+const DisableSSL = true
+
+func initDb() error {
+	storage := Storage{
+		Table:         TestTableName,
+		AwsEndpoint:   os.Getenv("AWS_ENDPOINT"),
+		AwsRegion:     os.Getenv("AWS_DEFAULT_REGION"),
+		AwsDisableSSL: true,
+	}
+	sess, err := session.NewSession(&aws.Config{
+		Endpoint:   &storage.AwsEndpoint,
+		Region:     &storage.AwsRegion,
+		DisableSSL: &storage.AwsDisableSSL,
+	})
+	if err != nil {
+		return err
+	}
+
+	svc := dynamodb.New(sess)
+
+	// attempt to delete table in case already exists
+	deleteTable := &dynamodb.DeleteTableInput{
+		TableName: aws.String(storage.Table),
+	}
+	_, err = svc.DeleteTable(deleteTable)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case dynamodb.ErrCodeResourceInUseException:
+				return aerr
+			case dynamodb.ErrCodeResourceNotFoundException:
+				// this is fine
+			case dynamodb.ErrCodeLimitExceededException:
+				return aerr
+			case dynamodb.ErrCodeInternalServerError:
+				return aerr
+			default:
+				return aerr
+			}
+		} else {
+			return err
+		}
+	}
+
+	// create table
+	createTable := &dynamodb.CreateTableInput{
+		AttributeDefinitions: []*dynamodb.AttributeDefinition{
+			{
+				AttributeName: aws.String("PrimaryKey"),
+				AttributeType: aws.String("S"),
+			},
+		},
+		KeySchema: []*dynamodb.KeySchemaElement{
+			{
+				AttributeName: aws.String("PrimaryKey"),
+				KeyType:       aws.String("HASH"),
+			},
+		},
+		ProvisionedThroughput: &dynamodb.ProvisionedThroughput{
+			ReadCapacityUnits:  aws.Int64(3),
+			WriteCapacityUnits: aws.Int64(3),
+		},
+		TableName: aws.String(storage.Table),
+	}
+	_, err = svc.CreateTable(createTable)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
```

Comment:
Are these all supposed to return the same thing? If so, is it worth having the `switch` statement?

Reply:
i'll clean it up